docs/docs/integrations/chat/llamacpp.ipynb

```diff
@@ -0,0 +1,448 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain.tools import tool\n",
+    "from langchain_core.pydantic_v1 import BaseModel, Field\n",
+    "\n",
+    "\n",
+    "class WeatherInput(BaseModel):\n",
+    "    location: str = Field(description=\"The city and state, e.g. San Francisco, CA\")\n",
+    "    unit: str = Field(enum=[\"celsius\", \"fahrenheit\"])\n",
+    "\n",
+    "\n",
+    "@tool(\"get_current_weather\", args_schema=WeatherInput)\n",
+    "def get_weather(location: str, unit: str):\n",
+    "    \"\"\"Get the current weather in a given location\"\"\"\n",
+    "    return f\"Now the weather in {location} is 22 {unit}\"\n",
+    "\n",
+    "\n",
+    "llm_with_tools = llm.bind_tools(\n",
+    "    tools=[get_weather],\n",
+    "    tool_choice={\"type\": \"function\", \"function\": {\"name\": \"get_current_weather\"}},\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Llama.generate: prefix-match hit\n",
+      "\n",
+      "llama_print_timings:        load time =    1178.78 ms\n",
+      "llama_print_timings:      sample time =     845.44 ms /    20 runs   (   42.27 ms per token,    23.66 tokens per second)\n",
+      "llama_print_timings: prompt eval time =     624.10 ms /    11 tokens (   56.74 ms per token,    17.63 tokens per second)\n",
+      "llama_print_timings:        eval time =    2785.06 ms /    19 runs   (  146.58 ms per token,     6.82 tokens per second)\n",
+      "llama_print_timings:       total time =    4400.99 ms /    30 tokens\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "AIMessage(content='', additional_kwargs={'function_call': {'name': 'get_current_weather', 'arguments': '{ \"location\": \"Ho Chi Minh City\", \"unit\" : \"celsius\"}'}, 'tool_calls': [{'index': 0, 'id': 'call__0_get_current_weather_cmpl-a6d145c7-358b-4e68-8c9e-49f3ee6802f6', 'type': 'function', 'function': {'name': 'get_current_weather', 'arguments': '{ \"location\": \"Ho Chi Minh City\", \"unit\" : \"celsius\"}'}}]}, response_metadata={'finish_reason': 'tool_calls'}, id='run-e6ec1cf7-299b-430c-bfa7-ac67322da3ec-0', tool_calls=[{'name': 'get_current_weather', 'args': {'location': 'Ho Chi Minh City', 'unit': 'celsius'}, 'id': 'call__0_get_current_weather_cmpl-a6d145c7-358b-4e68-8c9e-49f3ee6802f6'}])"
+      ]
+     },
+     "execution_count": 15,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "ai_msg = llm_with_tools.invoke(\n",
+    "    \"what is the weather like in HCMC in celsius\",\n",
+    ")\n",
+    "ai_msg"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "[{'name': 'get_current_weather',\n",
+       "  'args': {'location': 'Ho Chi Minh City', 'unit': 'celsius'},\n",
```

Comment:
Was unable to repro tool use working. 

Using:
```
* Weights: `meta-llama-3-8b-instruct.Q8_0.gguf`
* Link: https://huggingface.co/SanctumAI/Meta-Llama-3-8B-Instruct-GGUF
```

Following:
```
from langchain.tools import tool
from langchain_core.pydantic_v1 import BaseModel, Field

class WeatherInput(BaseModel):
    location: str = Field(description="The city and state, e.g. San Francisco, CA")
    unit: str = Field(enum=["celsius", "fahrenheit"])

@tool("get_current_weather", args_schema=WeatherInput)
def get_weather(location: str, unit: str):
    """Get the current weather in a given location"""
    return f"Now the weather in {location} is 22 {unit}"

llm_with_tools = llm.bind_tools(
    tools=[get_weather],
    tool_choice={"type": "function", "function": {"name": "get_current_weather"}},
)

ai_msg = llm_with_tools.invoke(
    "what is the weather in San Francisco, CA",
)
```

Raw output:
```
from_string grammar:
char ::= [^"\] | [\] char_1 
char_1 ::= ["\/bfnrt] | [u] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] 
location-kv ::= ["] [l] [o] [c] [a] [t] [i] [o] [n] ["] space [:] space string 
space ::= space_7 
string ::= ["] string_8 ["] space 
root ::= [{] space location-kv [,] space unit-kv [}] space 
unit-kv ::= ["] [u] [n] [i] [t] ["] space [:] space unit 
space_7 ::= [ ] | 
string_8 ::= char string_8 | 
unit ::= ["] [c] [e] [l] [s] [i] [u] [s] ["] | ["] [f] [a] [h] [r] [e] [n] [h] [e] [i] [t] ["] 

llama_tokenize_internal: Added a BOS token to the prompt as specified by the model but the prompt also starts with a BOS token. So now the final prompt starts with 2 BOS tokens. Are you sure this is what you want?
Llama.generate: prefix-match hit

llama_print_timings:        load time =   15285.72 ms
llama_print_timings:      sample time =     137.57 ms /    18 runs   (    7.64 ms per token,   130.84 tokens per second)
llama_print_timings: prompt eval time =     573.48 ms /    10 tokens (   57.35 ms per token,    17.44 tokens per second)
llama_print_timings:        eval time =    1963.75 ms /    17 runs   (  115.51 ms per token,     8.66 tokens per second)
llama_print_timings:       total time =    2744.26 ms /    27 tokens
```

Does not inoke tool as expected:
https://smith.langchain.com/public/2775cfe6-a138-49f3-9b7d-a791c0325246/r

Reply:
Will dig in further.

But, in any case, we may need more clear instructions such as:

1/ exact model to download (I believe I got same as you from above link)
2/ add `llama-cpp-python` or `llama.cpp` deps? i assume need to `make llama.cpp`

Reply:
I believe that have to work with any LLM model from huggingface

Reply:
Probably. It worth check on two models
Let's use gemma\llama2  as an alternative option